### ISSUE TITLE
comparison vs pomegranate

### ISSUE THREAD

**Issue by jmschrei** [hmmlearn/hmmlearn#104]
Hello again!

I wrote up this simple comparison vs pomegranate. I wanted to know if you thought it was fair to hmmlearn and if there are any other features you think I should include.

https://github.com/jmschrei/pomegranate/blob/master/benchmarks/pomegranate_vs_hmmlearn.ipynb

Thanks!

**Comment by soupault** [comment]
Waow! @jmschrei that's cool!

**Comment by superbobry** [comment]
Hi Jacob,

Thank you, this looks very interesting! I will have a good look at the notebook over the weekend, but I do have a couple of comments and questions right away.

* Am I right that "Structure Learning" correspond to learning the appropriate number of hidden states? If yes, then `hmmlearn` does not currently implement this.
* Could you please comment on the meaning of "Forward-Backward Transitions"? I've never encountered this term.
* `hmmlearn` is capable of "Multi-threaded Training", but this feature is not yet exposed to the user.

**Comment by jmschrei** [comment]
* Structure Learning: Okay, I'll rename it to be more accurate (which is that you don't need to define initial distributions and transition probabilities for your models)
* Forward-Backward Transitions: The forward-backward algorithm allows you to calculate the expected number of times a transition is used in addition to just the posterior estimates for the emissions. They can both be calculated at the same time.
* Multi-threaded Training: Are there any plans to implement this exposed to the user? I may remove it from the table if it's ambiguous.

**Comment by superbobry** [comment]
> The forward-backward algorithm allows you to calculate the expected number of times a transition is used in addition to just the posterior estimates for the emissions. They can both be calculated at the same time.

Yep, IIRC this quantity is needed for the EM-algorithm. Does pomegranate have an API method to compute it?

> Are there any plans to implement this exposed to the user? 

Sure, the current plan is to include it in the next release.

**Comment by jmschrei** [comment]
pomegranate does have an API call for it. forward_backward returns both matrices (while predict_proba returns only the emission one).

I'll remove the multi-threaded part for now then to not quibble over timing details then. Thanks for the comments so far!

**Comment by jmschrei** [comment]
Have you had a chance to look further into the comparison?

**Comment by chuston** [comment]
@jmschrei Does pomegranate only support one feature per HMM? It appears that each state can only have one emission distribution implying only one feature?

**Comment by chuston** [comment]
@jmschrei Ah, yes pomegranate does support multiple features through multivariate distributions. Nevermind.
